pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java

```diff
@@ -2166,8 +2169,35 @@ protected void processResults(ResultHandler handler, int flags) throws IOExcepti
           SimpleQuery currentQuery = executeData.query;
           Portal currentPortal = executeData.portal;
 
+          String nativeSql = currentQuery.getNativeQuery().nativeSql;
+          // Certain backend versions (e.g. 12.2, 11.7, 10.12, 9.6.17, 9.5.21, etc)
+          // silently rollback the transaction in the response to COMMIT statement
+          // in case the transaction has failed.
+          // See discussion in pgsql-hackers: https://www.postgresql.org/message-id/b9fb50dc-0f6e-15fb-6555-8ddb86f4aa71%40postgresfriends.org
+          if (isRaiseExceptionOnSilentRollback()
+              && handler.getException() == null
+              && status.startsWith("ROLLBACK")
+              && ("COMMIT".equalsIgnoreCase(nativeSql)
+              || COMMIT_PATTERN.matcher(nativeSql).find()
+              && !ROLLBACK_PATTERN.matcher(nativeSql).find())) {
```

Comment:
Indentation is funny, since the starting parenthesis in `&& ("COMMIT".eq...` is easily overlooked.

I would add another parenthesis around the pattern matching `COMMIT_PATTERN.matcher(nativeSql).find() && !ROLLBACK_PATTERN.matcher(nativeSql).find()` for clarity.

Is this pattern matching check really necessary? Since it does not consider quoting, comments or word boundaries, it would also be triggered if columns are named "committed_revision" or "rollback_reason"...

Reply:
>it would also be triggered if columns are named "committed_revision" or "rollback_reason"...

Can you imagine an SQL that would contain `commit_version` *column* and end up with `ROLLBACK` command status tag?

So far I know three SQL which might end up with `ROLLBACK`:

1) `ROLLBACK`
2) `COMMIT`
3) `PREPARE TRANSACTION`

So regexp here should be pretty much OK.

Reply:
I moved the condition to helper methods and added `prepare transaction` handling.